### DISCUSSION TITLE
How to prove certain partitions and files are not scanned

### DISCUSSION THREAD

**Discussion by SubieG** [duckdb/duckdb#5706] [Q&A]
Hi @samansmink , 
The hive parquet partitioning tests say this:
"This query should trigger the file skipping mechanism, which prevents reading metadata for files that are not scanned"
https://github.com/duckdb/duckdb/blob/77592d18a82413ac4901d990d127f0e3099ecae2/test/sql/copy/parquet/parquet_hive.test#L51

**How can it be proven that other partitions are not being read?**

With version `duckdb 0.6.1` we used the same data as the repo test, and the same query with the following:

`SET explain_output='all';EXPLAIN select id, value, date from parquet_scan('different_order/*/*/test.parquet', HIVE_PARTITIONING=1) where date = '2012-01-01';`

And we get a plan like this:
```
[('logical_plan',
  '┌───────────────────────────┐\n'
  '│         PROJECTION        │\n'
  '│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │\n'
  '│             id            │\n'
  '│           value           │\n'
  '│            date           │\n'
  '└─────────────┬─────────────┘                             \n'
  '┌─────────────┴─────────────┐\n'
  '│           FILTER          │\n'
  '│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │\n'
  "│   (date = '2012-01-01')   │\n"
  '└─────────────┬─────────────┘                             \n'
  '┌─────────────┴─────────────┐\n'
  '│        PARQUET_SCAN       │\n'
  '└───────────────────────────┘                             \n'),
 ('logical_opt',
  '┌───────────────────────────┐\n'
  '│         PROJECTION        │\n'
  '│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │\n'
  '│             id            │\n'
  '│           value           │\n'
  '│            date           │\n'
  '└─────────────┬─────────────┘                             \n'
  '┌─────────────┴─────────────┐\n'
  '│        PARQUET_SCAN       │\n'
  '└───────────────────────────┘                             \n'),
 ('physical_plan',
  '┌───────────────────────────┐\n'
  '│         PROJECTION        │\n'
  '│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │\n'
  '│             id            │\n'
  '│           value           │\n'
  '│            date           │\n'
  '└─────────────┬─────────────┘                             \n'
  '┌─────────────┴─────────────┐\n'
  '│        PARQUET_SCAN       │\n'
  '│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │\n'
  '│            date           │\n'
  '│             id            │\n'
  '│           value           │\n'
  '│            EC=1           │\n'
  '└───────────────────────────┘                             \n')]
```

1. Why doesn't the filter appear in the PARQUET_SCAN in the physical plan?
2. What does "EC=" mean in the physical plan?
3. Is the logical plan suggesting that all parquet files ARE scanned and THEN the filter is applied? (Again, doesn't this suggest all partitions were scanned--i.e. the file skipping mechanism didn't actually work?)

This is a tiny test, but we're trying to better understand when certain partitions are scanned or skipped before we scale up.
Thanks

**Comment by Tishj** [comment]
2. I believe EC stands for Estimated Cardinality

**Comment by samansmink** [comment]
@SubieG sorry for the late response here, i thought i had sent a message here

1. This is indeed a bit confusing atm, currently during pushdown we remove the filters that have been applied through the filenames (such as the hive partition columns). This results in them not showing up in the physical plan. Being able to see which files are actually scanned in a query (both metadata and data) is indeed super valuable and is probably something we should implement.
2. @Tishj 👍 
3. It does suggest that but that's not actually what's happening: DuckDB will create a list from the file system with all matches to the glob pattern`different_order/*/*/test.parquet`. Then it will analyse the filters and prune all files from this list that can never match the filter. Therefore these files will never be touched. There's one exception: the first file duckdb reads could be any file (the first file in the resulting list from the glob). This file is necessary for the bind phase so this metadata will be read anyways even though it might not match the filter through its hive partitions

hope that helps!
